RFC/RFC-14-p2p-adapter.md

```diff
@@ -0,0 +1,348 @@
+
+## Approach
+
+This approach is simple, it should be easy to debug and it could be refined in the future for performance if needed. 
+
+This implementation requires two extra services:
+
+- Messaging service: in charge of sending messages between specific peers.
+- Routing service: in charge of keeping track of the connections between peer and given them back routing information.
+
+As defined in [ADR-81](/ADR/ADR-81-minimum-comms-transport.md), each peer will known the ids of the peers around them using the information provided by the CommunicationsDirector.
+
+The basic idea of this implementation is for peers to connect randomly to each other forming a mesh, and reporting their connections to the routing service. The routing service will build routing tables for each peer to connect to every other. Each peer will receive from the routing service the path to all other conected peers. When a peer delivers a message, it will use the path provided by the routing table, and if there is no one, it will fallback to the messaging service.
+
+Although this implementation requires a server, it tries to maximize the usage of P2P connections, while also keeping in mind that peers may run in a constraint enviroment (like a browser) in which not always is possible to have a lot of open connections.
+
+## Flow example
+
+### 1. Peers establish random connections with their known peers.
+
+```mermaid
+sequenceDiagram
+    participant Peer1
+    participant Peer2
+    participant Peer3
+    participant Peer4
+    participant Peer5
+
+    Peer1->>Peer2: Initiate P2P webrtc connection
+    Peer2->>Peer1: Establish P2P webrtc connection
+
+    Peer1->>Peer3: Initiate P2P webrtc connection
+    Peer3->>Peer1: Establish P2P webrtc connection
+
+    Peer1->>Peer4: Initiate P2P webrtc connection
+    Peer4->>Peer1: Establish P2P webrtc connection
+
+    Peer2->>Peer3: Initiate P2P webrtc connection
+    Peer3->>Peer2: Establish P2P webrtc connection
+```
+
+### 2. Each peer report its connections to the routing service
+
+```mermaid
+sequenceDiagram
+    participant Peer1
+    participant Peer2
+    participant Peer3
+    participant Peer4
+    participant Peer5
+    participant RS as Routing Service
+
+    Peer1->>RS: connected to [peer2, peer3, peer4]
+    Peer2->>RS: connected to [peer1, peer3]
+    Peer3->>RS: connected to [peer1, peer2]
+    Peer4->>RS: connected to [peer1]
+    Peer5->>RS: connected to []
+```
+
+### 3. The routing service create a routing table for each peer
+
+A route is a list of peer ids. It doesn't contain the source or destination, so for example if a->b->c:
+- the route between a and c is [ b ]
+- and the route between b and c is []
+
+```mermaid
+sequenceDiagram
+    participant Peer1
+    participant Peer2
+    participant Peer3
+    participant Peer4
+    participant Peer5
+    participant RS as Routing Service
+
+    RS->>Peer1: { peer2: [], peer3: [], peer4: [] }
```

Comment:
So, RS never tells anyone about routes (or actually lack of routes) to `peer5`. So peers know about `peer5` because of archipelago/bff right?

Reply:
yes